erts/emulator/beam/erl_db_util.c

```diff
@@ -4030,54 +4030,119 @@ dmc_map(DMCContext *context, DMCHeap *heap, DMC_STACK_TYPE(UWord) *text,
         return retOk;
     } else {
         DECLARE_WSTACK(wstack);
+        DMC_STACK_TYPE(UWord) instr_save;
         Eterm *kv;
         int c;
         int textpos = DMC_STACK_NUM(*text);
-        int stackpos = context->stack_used;
+        int preventive_bumps = 0;
 
         ASSERT(is_hashmap(t));
 
         hashmap_iterator_init(&wstack, t, 1);
         constant_values = 1;
         nelems = hashmap_size(t);
 
-        /* Check if all keys and values are constants */
+        /* Check if all keys and values are constants. We do preventive_bumps for
+           all constants we find so that if we find a non-constant, the stack
+           depth will be correct. */
         while ((kv=hashmap_iterator_prev(&wstack)) != NULL && constant_values) {
```

Comment:
Suggestion:
Remove `constant_values` from this while-loop and only use `c` (with a better name maybe). We break out at first sight of `!c` and `constant_values` is kind of misleading as it's used for both keys and values.